apps/events/models.py

```diff
@@ -416,43 +416,67 @@ def generate_external_id(name: str, experiment_id: int, participant_id: int, ins
         return get_next_unique_id(ScheduledMessage, inputs, "external_id", length=5, model_instance=instance)
 
-    def safe_trigger(self):
-        """This wraps a call to the _trigger method in a try-catch block"""
+    def safe_trigger(self, attempt_number=1):
+        """Wraps _trigger with attempt tracking and retry"""
+        from apps.events.tasks import retry_scheduled_message
+
+        trigger_number = self.total_triggers
+        attempt = ScheduledMessageAttempt(
+            scheduled_message=self, trigger_number=trigger_number, attempt_number=attempt_number, team=self.team
+        )
+
         try:
-            self._trigger()
+            trace_metadata = self._trigger()
+            attempt.attempt_result = EventLogStatusChoices.SUCCESS
+            attempt.trace_info = trace_metadata
+            attempt.save()
+
         except Exception as e:
             logger.exception(f"An error occurred while trying to send scheduled message {self.id}. Error: {e}")
+            trace_metadata = getattr(e, "trace_metadata", {})
+            attempt.attempt_result = EventLogStatusChoices.FAILURE
+            attempt.log_message = str(e)
+            attempt.trace_info = trace_metadata
+            attempt.save()
+            if attempt_number < 3:
+                backoff_seconds = 2 ** (attempt_number - 1)
+                retry_scheduled_message.apply_async(args=[self.id, attempt_number + 1], countdown=backoff_seconds)
 
     def _trigger(self):
         experiment_session = self.participant.get_latest_session(experiment=self.experiment)
         if not experiment_session:
             # Schedules probably created by the API
             return
-
         trace_info = TraceInfo(
             name="scheduled message",
             metadata={
                 "schedule_id": self.external_id,
                 "trigger_number": self.total_triggers,
             },
         )
-        experiment_session.ad_hoc_bot_message(
-            self.params["prompt_text"],
-            trace_info,
-            fail_silently=True,
-            use_experiment=self._get_experiment_to_generate_response(),
-        )
-
-        utc_now = timezone.now()
-        self.last_triggered_at = utc_now
-        self.total_triggers += 1
-        if self._should_mark_complete():
-            self.is_complete = True
-        else:
-            delta = relativedelta(**{self.params["time_period"]: self.params["frequency"]})
-            self.next_trigger_date = utc_now + delta
-
-        self.save()
+        trace_metadata = {}
+        try:
+            trace_metadata = experiment_session.ad_hoc_bot_message(
+                self.params["prompt_text"],
+                trace_info,
+                fail_silently=False,
+                use_experiment=self._get_experiment_to_generate_response(),
+            )
+        except Exception as e:
+            trace_metadata = getattr(e, "trace_metadata", {})
+            e.trace_metadata = trace_metadata
+            raise
```

Comment:
This can be removed since the exception already has the `trace_metadata` attribute set in `ad_hoc_bot_message`